nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+    /// message indicating that they had been started
+    inferred_done: BTreeSet<SagaId>,
+
+    /// sagas that may be done, but we can't tell yet.  These are sagas where we
+    /// previously had them running in this process and the database state now
+    /// says that they're not running, but the database snapshot was potentially
+    /// from before the time that the saga started, so we cannot tell yet
+    /// whether the saga finished or just started.  We'll be able to tell during
+    /// the next pass and if it's done at that point then these sagas will move
+    /// to `inferred_done`.
+    maybe_done: BTreeSet<SagaId>,
+}
+
+impl Plan {
+    /// For a given saga recovery pass, determine what to do with each found
+    /// saga
+    ///
+    /// This function accepts:
+    ///
+    /// * `rest_state`: the at-rest saga recovery state from the end of the
+    ///   previous pass
+    /// * `running_sagas_found`: a list of sagas that the database reports
+    ///   to be in progress
+    ///
+    /// It determines:
+    ///
+    /// * which in-progresss sagas we don't need to do anything about because
+    ///   they're already running in this process (those sagas that are in both
+    ///   `sagas_started` and `running_sagas_found`)
+    /// * which sagas need to be recovered (those sagas in `running_sagas_found`
+    ///   but not in `sagas_started`)
+    /// * which sagas can be removed from `sagas_started` because they have
+    ///   finished (those in `previously_maybe_done` and *not* in
+    ///   `running_sagas_found`)
+    pub fn new(
+        log: &slog::Logger,
+        rest_state: &RestState,
+        mut running_sagas_found: BTreeMap<SagaId, nexus_db_model::Saga>,
+    ) -> Plan {
+        let mut builder = PlanBuilder::new(log);
+
+        // First of all, remove finished sagas from our "ignore" set.
+        //
+        // `previously_maybe_done` was computed the last time we ran and
+        // contains sagas that either just started or already finished.  We
+        // couldn't tell.  All we knew is that they were running in-memory but
+        // were not included in our database query for in-progress sagas.  At
+        // this point, though, we've done a second database query for
+        // in-progress sagas.  Any items that aren't in that list either cannot
+        // still be running, so we can safely remove them from our ignore set.
+        let previously_maybe_done = &rest_state.remove_next;
+        for saga_id in previously_maybe_done {
+            if !running_sagas_found.contains_key(saga_id) {
+                builder.saga_infer_done(*saga_id);
+            }
+        }
+
+        // Figure out which of the candidate sagas can clearly be skipped.
+        // Correctness here requires that the caller has already updated the set
+        // of sagas that we're ignoring to include any that may have been
+        // created up to the beginning of the database query.  (They do that by
```

Comment:
You can mostly ignore this comment. I'm just writing it down for lack of a better place and to see if maybe it will pique your interest / spark an idea.

I've read your comments in saga background task and most of that logic, but I'm still a bit concerned that correctness in this data structure requires such specific ordering in the background saga. However, it all appears correct and the excellent testing shows so. I'm wondering if there's a better way to enforce this ordering requirement in code so that it doesn't rely on comments to remain correct. I think one way to do that would be to use some sort of typestate process. But given that if the db query fails, we proceed anyway in the background task, this may add too much complexity in terms of states to be worth it.